// node-pushnotifications/lib/sendGCM.js
"use strict";

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
const gcm = require('node-gcm');
const R = require('ramda');
const _require = require('./constants'),
  DEFAULT_TTL = _require.DEFAULT_TTL,
  GCM_METHOD = _require.GCM_METHOD,
  GCM_MAX_TTL = _require.GCM_MAX_TTL;
const ttlFromExpiry = R.compose(R.min(GCM_MAX_TTL), expiry => expiry - Math.floor(Date.now() / 1000));
const extractTimeToLive = R.cond([[R.propIs(Number, 'expiry'), ({
  expiry
}) => ttlFromExpiry(expiry)], [R.propIs(Number, 'timeToLive'), R.prop('timeToLive')], [R.T, R.always(DEFAULT_TTL)]]);
const pathIsString = R.pathSatisfies(R.is(String));
const containsValidRecipients = R.either(pathIsString(['recipients', 'to']), pathIsString(['recipients', 'condition']));
const propValueToSingletonArray = propName => R.compose(R.of, R.prop(propName));
const getRecipientList = R.cond([[R.has('registrationTokens'), R.prop('registrationTokens')], [R.has('to'), propValueToSingletonArray('to')], [R.has('condition'), propValueToSingletonArray('condition')]]);
const sendChunk = (GCMSender, recipients, message, retries) => new Promise(resolve => {
  const recipientList = getRecipientList(recipients);
  GCMSender.send(message, recipients, retries, (err, response) => {
    // Response: see https://developers.google.com/cloud-messaging/http-server-ref#table5
    if (err) {
      resolve({
        method: GCM_METHOD,
        success: 0,
        failure: recipientList.length,
        message: recipientList.map(value => ({
          originalRegId: value,
          regId: value,
          error: err,
          errorMsg: err instanceof Error ? err.message : err
        }))
      });
    } else if (response && response.results !== undefined) {
      let regIndex = 0;
      resolve({
        method: GCM_METHOD,
        multicastId: response.multicast_id,
        success: response.success,
        failure: response.failure,
        message: response.results.map(value => {
          const regToken = recipientList[regIndex];
          regIndex += 1;
          return {
            messageId: value.message_id,
            originalRegId: regToken,
            regId: value.registration_id || regToken,
            error: value.error ? new Error(value.error) : null,
            errorMsg: value.error ? value.error.message || value.error : null
          };
        })
      });
    } else {
      resolve({
        method: GCM_METHOD,
        multicastId: response.multicast_id,
        success: response.success,
        failure: response.failure,
        message: recipientList.map(value => ({
          originalRegId: value,
          regId: value,
          error: new Error('unknown'),
          errorMsg: 'unknown'
        }))
      });
    }
  });
});
const sendGCM = (regIds, data, settings) => {
  const opts = _objectSpread({}, settings.gcm);
  const id = opts.id;
  delete opts.id;
  const GCMSender = new gcm.Sender(id, opts);
  const promises = [];
  const notification = {
    title: data.title,
    // Android, iOS (Watch)
    body: data.body,
    // Android, iOS
    icon: data.icon,
    // Android
    image: data.image,
    // Android
    picture: data.picture,
    // Android
    style: data.style,
    // Android
    sound: data.sound,
    // Android, iOS
    badge: data.badge,
    // iOS
    tag: data.tag,
    // Android
    color: data.color,
    // Android
    click_action: data.clickAction || data.category,
    // Android, iOS
    body_loc_key: data.locKey,
    // Android, iOS
    body_loc_args: data.locArgs,
    // Android, iOS
    title_loc_key: data.titleLocKey,
    // Android, iOS
    title_loc_args: data.titleLocArgs,
    // Android, iOS
    android_channel_id: data.android_channel_id,
    // Android
    notification_count: data.notificationCount || data.badge // Android
  };
  let custom;
  if (typeof data.custom === 'string') {
    custom = {
      message: data.custom
    };
  } else if (typeof data.custom === 'object') {
    custom = _objectSpread({}, data.custom);
  } else {
    custom = {
      data: data.custom
    };
  }
  custom.title = custom.title || data.title;
  custom.message = custom.message || data.body;
  custom.sound = custom.sound || data.sound;
  custom.icon = custom.icon || data.icon;
  custom.msgcnt = custom.msgcnt || data.badge;
  if (opts.phonegap === true && data.contentAvailable) {
    custom['content-available'] = 1;
  }
  const message = new gcm.Message({
    // See https://developers.google.com/cloud-messaging/http-server-ref#table5
    collapseKey: data.collapseKey,
    priority: data.priority === 'normal' ? 'normal' : 'high',
    contentAvailable: data.silent ? true : data.contentAvailable || false,
    delayWhileIdle: data.delayWhileIdle || false,
    timeToLive: extractTimeToLive(data),
    restrictedPackageName: data.restrictedPackageName,
    dryRun: data.dryRun || false,
    data: opts.phonegap === true ? Object.assign(custom, notification) : custom,
    // See https://github.com/phonegap/phonegap-plugin-push/blob/master/docs/PAYLOAD.md#android-behaviour
    notification: opts.phonegap === true || data.silent === true ? undefined : notification
  });
  let chunk = 0;

  /* allow to override device tokens with custom `to` or `condition` field:
   * https://github.com/ToothlessGear/node-gcm#recipients */
  if (containsValidRecipients(data)) {
    promises.push(sendChunk(GCMSender, data.recipients, message, data.retries || 0));
  } else {
    // Split tokens in 1.000 chunks, see https://developers.google.com/cloud-messaging/http-server-ref#table1
    do {
      const registrationTokens = regIds.slice(chunk * 1000, (chunk + 1) * 1000);
      promises.push(sendChunk(GCMSender, {
        registrationTokens
      }, message, data.retries || 0));
      chunk += 1;
    } while (1000 * chunk < regIds.length);
  }
  return Promise.all(promises).then(results => {
    const resumed = {
      method: GCM_METHOD,
      multicastId: [],
      success: 0,
      failure: 0,
      message: []
    };
    results.forEach(result => {
      if (result.multicastId) {
        resumed.multicastId.push(result.multicastId);
      }
      resumed.success += result.success;
      resumed.failure += result.failure;
      resumed.message.push(...result.message);
    });
    return resumed;
  });
};
module.exports = sendGCM;
